### ISSUE TITLE
Wayland users, how does your window manager identify itself?

### ISSUE THREAD

**Issue by grtcdr** [Macchina-CLI/libmacchina#90]
Wayland currently has no central place where we can fetch the name of the window manager that you use.

My question is, how does your window manager identify itself?

In the case of Sway, it sets an environment variable `SWAYSOCK` that points to a socket on the filesystem, which we can quickly and easily check.

I don't know about the others, but maybe you can tell us.

**Comment by ravenclaw900** [comment]
Default GNOME on Fedora 36 fails with `Unknown window manager`.

**Comment by grtcdr** [comment]
Do `XDG_CURRENT_DESKTOP` or `DESKTOP_SESSION` say anything?

**Comment by ravenclaw900** [comment]
`GNOME` and `gnome`, respectively. I got the window manager by the fact that it was using a socket (shell command: `ps -p $(lsof -t "$XDG_RUNTIME_DIR/$WAYLAND_DISPLAY") -ho comm`). I get `gnome-shell`, which is a Mutter plugin.

**Comment by grtcdr** [comment]
Yeah, we don't currently look for mutter, as can be seen in [this file](https://github.com/Macchina-CLI/libmacchina/blob/main/src/winman.rs).

Because there's no centralized place to query for this information, we sort of have to resort to detection in the most naive way possible, which currently doesn't support nested wayland compositors.

**Comment by grtcdr** [comment]
[wayland-info](https://gitlab.freedesktop.org/ofourdan/wayland-info) reports lots of data, but none of it mentions the name of my compositor, which leads me to believe that there's currently no way, perhaps in the future though.

**Comment by Absolpega** [comment]
i would try and copy fastfetch
they get the DE from `$XDG_CURRENT_DESKTOP`

the WM is a little more complicated
fastfetch seems to be following this stackoverflow answer https://stackoverflow.com/a/72755046
it gets a file descriptor using wl_display_get_fd
then gets the pid of the process connected to the socket
and then gets the name

**Comment by Absolpega** [comment]
i actually just wrote an implementation for my fetch that uses libmacchina
```rust
pub fn window_manager() -> Error<String> {
    if !is_lib_available() {
        return Err(ReadoutError::MetricNotAvailable);
    }

    let display_ptr = unsafe {
        ffi_dispatch!(
            wayland_client_handle(),
            wl_display_connect,
            ::std::ptr::null()
        )
    };

    if display_ptr.is_null() {
        return Err(ReadoutError::MetricNotAvailable);
    }

    let display_fd =
        unsafe { ffi_dispatch!(wayland_client_handle(), wl_display_get_fd, display_ptr) }
            .as_raw_fd();

    let pid = getsockopt(display_fd, sockopt::PeerCredentials)
        .map_err(|_| ReadoutError::MetricNotAvailable)?
        .pid();

    Ok(std::fs::read_to_string(format!("/proc/{}/comm", pid))?)
}
```

**Comment by Absolpega** [comment]
would be my first contribution to an open source project
always happy to learn
yea, i can try

**Comment by Absolpega** [comment]
i made the pull request
